utils/emmylua/wesnoth.lua

```diff
@@ -0,0 +1,272 @@
+---@param type '"markov"' The type of generator to create
+---@param definition string[] Sample names to be input to the generator
+---@param chain_size? integer The size of the Markov chain unit
+---@param max_length? integer Maximum length of the generated names
+---@return fun():string
+function wesnoth.name_generator(type, definition, chain_size, max_length) end
+
+---Compile a WFL formula into a Lua function
+---@param formula string A WFL formula
+---@return fun(variables:WML):any
+function wesnoth.compile_formula(formula, variables) end
+
+---Evaluate a WFL formula
+---@param formula string A WFL formula
+---@param variables WML Table defining WFL variables.
+---@return any
+function wesnoth.eval_formula(formula, variables) end
+
+---Represents an arbitrary version number.
+---@class version
+---@field major integer The major revision number (index 1)
+---@field minor integer The minor revision number (index 2)
+---@field revision integer The patch revision number (index 3)
+---@field is_canonical boolean Whether the version has at most three components
+---@field sep string|nil The character separating the suffix from the main version
+---@field special string The version suffix
+
+---Construct a version
+---@param str string A version string to parse
+---@param major integer The major revision number
+---@param minor? integer The minor revision number
+---@param revision? integer The patch revision number
+---@param suffix string The suffix, including the separator character if any
+---@return version
+---@overload fun(major:integer, minor?:integer, revision?:integer, suffix?:string):version
+function wesnoth.version(str) end
+
+---@return version #The current version of Wesnoth
+function wesnoth.current_version() end
+
+---@return integer #The number of milliseconds since Wesnoth launched
+function wesnoth.ms_since_init() end
+
+---Output a deprecated message
+---@param element_name string The name of the element being deprecated
+---@param level '1'|'2'|'3'|'4' The deprecation level
+---@param version string|nil The earliest version the element may be removed in
+---@param detail_message An additional message describing the deprecation and usually indicating a replacement
+function wesnoth.deprecated_message(element_name, level, version, detail_message) end
+
+---@type table<string, fun(cfg:WML)>
+wesnoth.wml_actions = {}
+
+---@type table<string, fun(cfg:WML):boolean>
+wesnoth.wml_conditionals = {}
+
+---@type table<string, fun(unit:unit, cfg:WML)>
+wesnoth.effects = {}
+
+---@type table<string, fun(cfg:WML)>
+wesnoth.custom_synced_commands = {}
+
+---@class color
+---@field r integer
+---@field g integer
+---@field b integer
+---@field a integer
+---@class color_range
+---@field mid color
+---@field min color
+---@field max color
+---@field minimap color
+---@field pango_color string
+
+---@type table<string, color_range>
+wesnoth.colors = {}
+
+---@class event_context
+---@field name string
```

Comment:
tstring ?

Reply:
Pretty sure this is the event name key from the WML, so not translatable.